### ISSUE TITLE
InputTake is broken for curstom input / slices

### ISSUE THREAD

**Issue by dathinab** [Geal/nom#535]
Due to some circumstances I have to use a custom Slice/Input type 
(roughly  `struct Slice<'a>(&'a str, usize)` ) and noticed that there is 
no good way to implement `InputTake` for it.

The current definition of input take is:
```rust
/// abstracts slicing operations
pub trait InputTake {
    /// returns a slice of `count` bytes
    fn take<P>(&self, count: usize)  -> Option<&Self>;
    /// split the stream at the `count` byte offset
    fn take_split<P>(&self, count: usize) -> Option<(&Self,&Self)>;
}

impl InputTake for  str { [...] }
[...]
```

The problem is that if you have a custom input type/slice the reference becomes part
of your type, but in InputTake it is returning  `&Self`'s which works for slices as they
know about all the fat-pointer compiler magic but cant work for custom types which
do not consist of just a fat reference (pointer+len) but a fat reference+somthing
(while normally still being Copy).

This can be solved by implementing `InputTake for &'a str` and changing it's return values
from `&Self` to  `Self`.

Also, why is there a `<P>` generic parameter??

As a side not this potentially also affects other traits
(e.g. `Offset`, but there it is not a problem, just a unnecessary layer of indirection)

**Comment by Geal** [comment]
I just ran into this problem while playing with `AtEof`, will fix it soon

**Comment by Geal** [comment]
Fixed in 784b1f9
